### ISSUE TITLE
[Question] Transfer SGX enclave from a server to another one

### ISSUE THREAD

**Issue by ThibsG** [oscarlab/graphene#1050]
Hi !

I want to know if there is a way to transfer a built SGX enclave to another server.
Both servers are Graphene and SGX ready.

I built python test example, and tried to send some of the files (`python.manifest.sgx`, `python.sig`, `python.token`) to the other server, in order to have _server1_ base enclave files on _server2_.

Though, when starting the enclave, i have the following error:
`initializing enclave failed: init_enclave: 1`

The goal is to share a same `MRENCLAVE` key across several servers.

Is that possible ? Am i missing files ?

Thanks !

**Comment by ThibsG** [comment]
If i understand the correct workflow about enclave transfer...

Server1
- build the enclave (i.e. `make SGX=1`). It creates python.{sgx, manifest.sgx, manifest.sgx.d}
- ship the enclave (=created files) to another server

Server2
- build the runtime tokens with the transferred enclave (i.e. `make SGX=1 sgx-tokens`)

Note that using this workflow, it outputs no error anymore.

Is that a correct way to share enclave across multiple servers ?

**Comment by dimakuv** [comment]
Your second message describes the correct way of moving enclaves from the developer machine (Server1) to a production machine (Server2).

The reason is that the EINITTOKEN must be generated by the Launch Enclave residing **on the same** machine as the enclave-to-run (at least in SGX version 1). You may find more information about it in "Intel SGX Explained", section 5.9. You can also look at the EINIT instruction of SGX.

**Comment by ThibsG** [comment]
That makes sense, thanks !

But i'm surprised i don't get the same Seal keys by running the shipped enclave on 2 different machines. Should not the MRENCLAVE Seal keys be the same on both machines ?

**Comment by mkow** [comment]
Sealing keys are "local" to the CPU you generate them on and will be different on different machine. MRENCLAVE is not the only input to the key derivation.
The intended usage for sealing keys is for storage encryption.

**Comment by ThibsG** [comment]
Well, of course sorry, i meant `MRSIGNER` (not `MRENCLAVE`).
Are Seal keys for `MRSIGNER` only local too ? 

As i understand from [Intel documentation](https://software.intel.com/en-us/node/702996), the `MRSIGNER` is tied to the built enclave (public RSA key).
So if we run the **same** enclave in machine A and machine B, i thought we might be able to get an `MRSIGNER` key which is the same between both machines.

Do you confirm is that possible in SGX ? If yes, is that still possible through Graphene ?
By the way, what is the role of the following secret : `Pal/src/host/Linux-SGX/signer/enclave-key.pem` ?

**Comment by mkow** [comment]
> Are Seal keys for `MRSIGNER` only local too ?

yes

> As i understand from Intel documentation, the MRSIGNER is tied to the built enclave (public RSA key).
> So if we run the same enclave in machine A and machine B, i thought we might be able to get an MRSIGNER key which is the same between both machines.

This isn't possible and I don't think it would make much sense to have such a feature in SGX. You can just deploy such a key after remote attestation to each enclave.

>  By the way, what is the role of the following secret : `Pal/src/host/Linux-SGX/signer/enclave-key.pem` ?

It's used to sign test binaries.

**Comment by ThibsG** [comment]
> This isn't possible and I don't think it would make much sense to have such a feature in SGX.

The idea behind this was to be sure code deployed in production was not modified.
How do we know the code from enclave A is still the same as the one running in enclave B ?

**Comment by mkow** [comment]
Through remote attestation, but this is only for the initial state, measuring the state after the enclave started execution doesn't make too much sense.
Please read some introduction document for SGX, e.g. the one mentioned by Dmitrii above. Also, please note that we are just a framework using SGX, we aren't responsible for SGX itself.

**Comment by dimakuv** [comment]
@ThibsG You may want to look at Figure 78 in "Intel SGX Explained". It shows the key derivation material. Notice `SEAL_FUSES` as one of them -- this is what ties the derived seal key to the local machine.

Shameless plug: https://www.youtube.com/watch?v=4YuTKl2ti4E&t=2s. This video tries to cover the whole story of "how porting apps to Intel SGX works".

**Comment by bgrieder** [comment]
@mkow @dimakuv another way of looking at it is: how do you guarantee that the content of the Graphene manifest file - and the referenced librairies- have not been changed when you moved to production?

**Comment by bgrieder** [comment]
@mkow @dimakuv To try to explain the case better, here is a sample scenario using Redis as in your video.

Player A prepares en enclave with a particular version of the Redis code, all of its dependencies, a Security Code and a Graphene `redis.manifest` file that it makes into a `redis.manifest.sgx`. The Security Code is some Player A custom code running n the enclave which is able to access all enclave secrets, produce quotes, reports and is able to access the file system. 
Player A can run this enclave and call the Security Code.

Player A forwards all the enclave files to Player B.

Player B runs `make SGX=1 sgx-tokens` and starts the enclave provided by Player A.

Before pushing data to player B Redis enclave, Player A wants to make sure Player B is running the exact same version of the Reids code that was shipped to B, and not some some other software that will pretend being Redis but will actually reveal any data pushed to it.

What can the Security Code running in Player B enclave provide as a proof that it is running the exact same enclave code as the one provided by Player A ?

nb: I work with @ThibsG

**Comment by dimakuv** [comment]
@bgrieder @ThibsG Here are the explanations. But first, you may want to take a look at the Secret Provisioning example here: https://github.com/thomasknauth/sgx-ra-tls/tree/provision-example/apps/redis-secrect-provisioning-example. This example also uses Redis, and the verifier (Player A) checks the enclave's MRENCLAVE + MRSIGNER measurements received from Player B as part of the SGX remote attestation.

> How does the enclave guarantee that the content of its Graphene manifest file - and the referenced librairies- have not been changed when you moved to production?

Graphene-SGX puts the manifest file in the measured enclave memory (see `https://github.com/oscarlab/graphene/blob/master/Pal/src/host/Linux-SGX/signer/pal-sgx-sign: generate_measurement()`). So, the manifest file is reflected in MRENCLAVE. And since the manifest file contains the hashes of all referenced trusted libraries, by induction all referenced libraries are reflected in MRENCLAVE. Now, (1) if the attacker modifies the manifest contents, then the run-time MRENCLAVE calculated by Intel SGX will not match the MRENCLAVE expected by Player 1. (2) If the attacker replaces one of the libraries, its run-time hash calculated by Graphene-SGX as part of the dynamic-load process will not match the predefined hash in the manifest.

Think of it as a chain of trust: `trust MRENCLAVE -> trust manifest contents -> trust referenced libraries`.

> What can the Security Code running in Player B enclave provide as a proof that it is running the exact same enclave code as the one provided by Player A ?

The remote attestation process of Intel SGX is used for this scenario. Basically, these steps:
1. Player A challenges Player B (with some nonce) saying "Player B, send me the attestation report that you are what you claim you are".
2. Player A waits for the report. Player A expects the report to contain MRENCLAVE/MRSIGNER/other attributes to be exactly as he expects.
3. Player B creates a Quote request in the form of the SGX report (with the nonce provided by A) and performs SGX local attestation with the special Quoting Enclave (Intel PSW enclave accessed through the AESM linux daemon). The Quoting Enclave generates the Local Quote.
4. The Local Quote must be verified by Intel IAS service, so Player B sends the Local Quote via HTTPS to Intel IAS.
5. Intel IAS verifies the authenticity of the Local Quote (IAS trusts PSW which signed this Local Quote + IAS knows about your identity when you registered for the SPID/key at the Intel web-site).
6. If IAS is happy with the Local Quote, it sends back the correct Attestation Report. This Attestation Report contains something like "I am Intel IAS, here is my certificate chain, I verified and vouch for the Local Quote sent to me by this particular enclave, and here are the measurements of the enclave: MRENCLAVE, MRSIGNER, nonce that someone put there, ...".
7. Player B receives this Attestation Report and forwards it to Player A.
8. Now Player A gets the report and verifies it. Intel IAS publishes and maintains its certificate chain, so Player A obtained these certificates off-band beforehand. Thus, Player A can verify the certificate chain (and thus assure that Intel vouches for the received report). Next Player A thinks "Ok, so this report is trusted by Intel, and Intel knows about its SGX enclaves, so I can trust the measurements of the enclave enclosed in this report". So Player A finally verifies MRENCLAVE, MRSIGNER, and the nonce embedded in this report. If they match to what Player A expects, **Player A finally trusts Player B's enclave**!

This is also (very similarly) explained in our code: https://github.com/oscarlab/graphene/blob/master/Pal/src/host/Linux-SGX/enclave_platform.c. Hope this helps.

**Comment by dimakuv** [comment]
We should start filling up Graphene FAQ with such questions and answers :)

**Comment by bgrieder** [comment]
@dimakuv Thank you very much for your detailed answer and yes, we agree, "standard" use cases like this one  - deploying to a remote infra such as a cloud infra - should go into an FAQ.

Yes, the right combination seems to be:
  - checking the MRENCLAVE -> to ensure code integrity
  - checking the remote attestation  -> to make sure the enclave is run on an SGX machine, not some simulated environment which would allow the attacker to dump memory.
 - going through IAS -> to better protect the target enclave identity.
  - using a nonce -> to randomize the process and minimize replay attacks.

Once thing we noticed though is that:
1. the enclave files cannot be run "as is" on the target machine (player B) unless `make SGX=1 sgx-tokens` is run on that machine
2. doing so changes the MRSIGNER of the enclave run on the target machine

We are not yet sure what the security implications of changing the MRSIGNER are in this use case. Anyway, we will have our in-house cryptographers go through the complete flow; we will post the result in this issue when we have reached conclusions (which may take some time).

In the meantime, thank you very much again for your quick and detailed answers which really help putting us on track to a robust solution. Feel free to close this issue before we post conclusions.

**Comment by dimakuv** [comment]
> 2. doing so changes the MRSIGNER of the enclave run on the target machine

Oh, this doesn't sound right. MRSIGNER must not change during `make SGX=1 sgx-tokens`, which should create a token (EINITTOKEN) based on the sig file (SIGSTRUCT).

One thing that I didn't articulate in my answer is that the developer must copy the `*.manifest.sgx` file and `*.sig` file. The latter one is created together with the manifest file during the normal build process (`make SGX=1`). The `*.sig` file is then what is used as the input for `make SGX=1 sgx-tokens`.

I can imagine that copying only the manifest file *without* the sig file will force Graphene to incorrectly (1) create the sig file using another, non-developer-machine key, and (2) create the token using this incorrect sig file.

You can also check Figure 85 of "Intel SGX Explained".

So, in short, one should do the following:
1. Build the following files on the developer machine using `make SGX=1`:
    - app binary and all dependent libraries
    - auto-generate the `*.manifest.sgx` file
    - auto-generate the `*.sig` file
2. Copy these files to the production machines and on each machine:
    - run `make SGX=1 sgx-tokens`
    - this will auto-generate `*.token` file
    - all `*.token` files must have the same MRSIGNER (derived from the MODULUS field in SIGSTRUCT stored in the `*.sig` file).

**Comment by ThibsG** [comment]
>     2. Copy these files to the production machines and on each machine:
>        
>        * run `make SGX=1 sgx-tokens`
>        * this will auto-generate `*.token` file
>        * all `*.token` files must have the same MRSIGNER (derived from the MODULUS field in SIGSTRUCT stored in the `*.sig` file).

Transferred `*.manifest`, `*.manifest.sgx` and `*.sig ` files to production machines.

**Note** that if i'm not transferring `*.manifest` file, `make` can't find rules to build `*.manifest` when running `make SGX=1 sgx-tokens` and stops.

When running `make SGX=1 sgx-tokens` in both machines, `*.sig` file is different between them.

**Comment by dimakuv** [comment]
> When running `make SGX=1 sgx-tokens` in both machines, `*.sig` file is different between them and thus the MRSIGNER is different.

Do I understand correctly that the `*.sig` file is overwritten when you do `make SGX=1 sgx-tokens`? This is not supposed to happen. (@mkow @yamahata @HW42, is this because copying files from one machine to another confuses make's logic on whether to rebuild `*.sig`?)

Anyway, instead of using `make SGX=1 sgx-tokens`, you can directly use:
```sh
<graphene>/Pal/src/host/Linux-SGX/signer/pal-sgx-get-token -output myapp.token -sig myapp.sig
# must generate myapp.token and NOT modify myapp.sig
```

**Comment by yamahata** [comment]
re-make can happen because copyed files will have newer time stamp. It depends on the order of copy, which file is copied first.

**Comment by ThibsG** [comment]
Your solution based on `src/host/Linux-SGX/signer/pal-sgx-get-token -output myapp.token -sig myapp.sig` is working perfectly.

I found a workaround using `make SGX=1 sgx-tokens`:
- send `*.manifest` and `*.manifest.sgx` from machine1 to machine2
- run `make SGX=1 sgx-tokens` on machine2
  - it generates a `*.sig` file (different from machine1)
  - at this step, only MRENCLAVE is the same between both machines
- send `*.sig` from machine1 to machine2
- run **again** `make SGX=1 sgx-tokens` to regenerate a valid `*.token` file
  - now MRENCLAVE and MRSIGNER are the same between both machines, 
which is what we expected !

I tried to send `*.sig` file with `*.manifest` and `*.manifest.sgx` (at step1), sending them in several orders to get different timestamps between them, but whatever the order, `*.sig` file is overwritten everytime.

Thank your for your patience, I can make further tests if it helps to debug Graphene, just let me know !

**Comment by dimakuv** [comment]
Thank you for this info! Glad you resolved your issue (with a workaround, but still). We'll look into make dependencies and ordering, something is not right there.

**Comment by ThibsG** [comment]
Thanks again, i'm now closing this issue as it is solved, feel free to open it again if you need us to do more tests.
